File: src/dolmir/orchestration/trace/epistemic.py
"""Epistemic primitives: evidence, claims, and their status tags.

Cognitive Constitution §8: facts, observations, assumptions, and hypotheses
are never conflated — every claim in DOLMIR's reasoning carries an explicit
epistemic status, and nothing downstream reads a claim without knowing
which one it is.

Cognitive Constitution §2 (grounding discipline) is enforced *structurally*
here, not by convention: a ``Claim`` tagged ``FACT`` cannot be constructed
without at least one citation or computation grounding it. The reasoning
data model has no way to represent an ungrounded fact.
"""

from __future__ import annotations

import enum
from dataclasses import dataclass, field

__all__ = ["Claim", "EpistemicStatus", "Evidence", "EvidenceKind"]


class EpistemicStatus(enum.Enum):
    """What kind of knowledge a claim represents (CC §8)."""

    FACT = "fact"
    """Grounded in cited doctrine or deterministic computation; not up for debate."""

    OBSERVATION = "observation"
    """A direct, low-inference reading of data."""

    ASSUMPTION = "assumption"
    """An interpretation layered on observations — labeled, never silently
    promoted to fact."""

    HYPOTHESIS = "hypothesis"
    """A forward-looking, inherently probabilistic claim."""


class EvidenceKind(enum.Enum):
    """Where a piece of evidence draws its authority from (CC §2)."""

    CITATION = "citation"
    """Retrieved from a curated source, referenced by stable id + version."""

    COMPUTATION = "computation"
    """Produced by a named deterministic computation over data in the system."""

    OBSERVATION = "observation"
    """A direct reading of input data available to this reasoning run."""


@dataclass(frozen=True, kw_only=True, slots=True)
class Evidence:
    """One grounded piece of support for a claim or stance.

    ``source_ref`` identifies the authority precisely enough to audit later:
    a document id + version for citations, a computation name for computed
    values, an input artifact reference for observations.
    """

    kind: EvidenceKind
    source_ref: str
    content: str

    def __post_init__(self) -> None:
        """Reject evidence that cannot be traced back to its source."""
        if not self.source_ref.strip():
            msg = "Evidence.source_ref must be non-empty — untraceable evidence is not evidence"
            raise ValueError(msg)
        if not self.content.strip():
            msg = "Evidence.content must be non-empty"
            raise ValueError(msg)


_FACT_GROUNDING_KINDS = frozenset({EvidenceKind.CITATION, EvidenceKind.COMPUTATION})


@dataclass(frozen=True, kw_only=True, slots=True)
class Claim:
    """A single statement with an explicit epistemic status and its grounding.

    Structural rules (all raise ``ValueError`` — the states are illegal, per
    Standing Rule 5):

    - ``FACT`` requires at least one ``CITATION`` or ``COMPUTATION`` evidence
      (CC §2: never asserted from parametric memory alone).
    - ``OBSERVATION`` requires at least one ``OBSERVATION`` evidence — a
      reading of data must point at the data it read.
    - ``ASSUMPTION`` and ``HYPOTHESIS`` may carry any evidence, including
      none: they are labeled interpretation, and the label is the point.
    """

    statement: str
    status: EpistemicStatus
    evidence: tuple[Evidence, ...] = field(default=())

    def __post_init__(self) -> None:
        """Enforce the grounding discipline for the declared status."""
        if not self.statement.strip():
            msg = "Claim.statement must be non-empty"
            raise ValueError(msg)
        if self.status is EpistemicStatus.FACT and not any(
            item.kind in _FACT_GROUNDING_KINDS for item in self.evidence
        ):
            msg = (
                "a FACT claim requires at least one CITATION or COMPUTATION "
                "evidence (Cognitive Constitution §2) — downgrade it to "
                "ASSUMPTION if no such grounding exists"
            )
            raise ValueError(msg)
        if self.status is EpistemicStatus.OBSERVATION and not any(
            item.kind is EvidenceKind.OBSERVATION for item in self.evidence
        ):
            msg = (
                "an OBSERVATION claim requires at least one OBSERVATION "
                "evidence pointing at the data it read"
            )
            raise ValueError(msg)

File: packages/core/src/kernel/epistemic.ts
import { z } from 'zod';

import { validationErrorFromZod, type ValidationError } from './errors.js';
import { err, ok, type Result } from './result.js';

/**
 * Epistemic primitives (ADR-0007, Directive §10–§11).
 *
 * Every claim DOLMIR makes carries an explicit status, and the grounding
 * discipline is structural: a FACT without computation, citation or a typed
 * record read cannot be constructed; an OBSERVATION must point at the data it
 * read. Untraceable evidence is rejected at construction.
 */

export const EpistemicStatus = {
  /** Grounded in a deterministic computation, a cited source or a typed record. Not up for debate. */
  FACT: 'FACT',
  /** A direct, low-inference reading of input data. */
  OBSERVATION: 'OBSERVATION',
  /** An interpretation layered on observations — labelled, never silently promoted. */
  ASSUMPTION: 'ASSUMPTION',
  /** A forward-looking, inherently uncertain claim. */
  HYPOTHESIS: 'HYPOTHESIS',
} as const;
export const EpistemicStatusSchema = z.enum(['FACT', 'OBSERVATION', 'ASSUMPTION', 'HYPOTHESIS']);
export type EpistemicStatus = z.infer<typeof EpistemicStatusSchema>;

export const EvidenceKind = {
  /** A curated or external source, referenced by stable id and version. */
  CITATION: 'CITATION',
  /** Produced by a named deterministic computation over data in the system. */
  COMPUTATION: 'COMPUTATION',
  /** A typed read of a persisted record field (table, id, field). */
  RECORD_FIELD: 'RECORD_FIELD',
  /** A verbatim span of an ingested document (document id, offsets or page). */
  DOCUMENT_SPAN: 'DOCUMENT_SPAN',
  /** A direct reading of an input available to this run. */
  OBSERVATION: 'OBSERVATION',
} as const;
export const EvidenceKindSchema = z.enum([
  'CITATION',
  'COMPUTATION',
  'RECORD_FIELD',
  'DOCUMENT_SPAN',
  'OBSERVATION',
]);
export type EvidenceKind = z.infer<typeof EvidenceKindSchema>;

const nonEmpty = (message: string) => z.string().trim().min(1, message);

export const EvidenceSchema = z
  .object({
    kind: EvidenceKindSchema,
    /** Precise enough to audit later: document id + span, computation name + inputs, table/id/field, source id + version. */
    sourceRef: nonEmpty(
      'Evidence.sourceRef must be non-empty — untraceable evidence is not evidence.',
    ),
    /** The supporting content itself (a quote, a computed value rendered, a record value). */
    content: nonEmpty('Evidence.content must be non-empty.'),
    /** Optional machine-readable locator (offsets, page, row id, field name…). */
    locator: z.record(z.string(), z.union([z.string(), z.number()])).optional(),
  })
  .strict();
export type Evidence = z.infer<typeof EvidenceSchema>;

const FACT_GROUNDING: ReadonlySet<EvidenceKind> = new Set([
  EvidenceKind.CITATION,
  EvidenceKind.COMPUTATION,
  EvidenceKind.RECORD_FIELD,
]);
const OBSERVATION_GROUNDING: ReadonlySet<EvidenceKind> = new Set([
  EvidenceKind.OBSERVATION,
  EvidenceKind.DOCUMENT_SPAN,
  EvidenceKind.RECORD_FIELD,
]);

export const ClaimSchema = z
  .object({
    statement: nonEmpty('Claim.statement must be non-empty.'),
    status: EpistemicStatusSchema,
    evidence: z.array(EvidenceSchema).default([]),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (
      value.status === EpistemicStatus.FACT &&
      !value.evidence.some((item) => FACT_GROUNDING.has(item.kind))
    ) {
      ctx.addIssue({
        code: 'custom',
        path: ['evidence'],
        message:
          'A FACT requires at least one CITATION, COMPUTATION or RECORD_FIELD evidence; downgrade it to ASSUMPTION if no such grounding exists.',
      });
    }
    if (
      value.status === EpistemicStatus.OBSERVATION &&
      !value.evidence.some((item) => OBSERVATION_GROUNDING.has(item.kind))
    ) {
      ctx.addIssue({
        code: 'custom',
        path: ['evidence'],
        message:
          'An OBSERVATION requires at least one OBSERVATION, DOCUMENT_SPAN or RECORD_FIELD evidence pointing at the data it read.',
      });
    }
  });
export type Claim = z.infer<typeof ClaimSchema>;
export type ClaimInput = z.input<typeof ClaimSchema>;

export function claim(input: ClaimInput): Result<Claim, ValidationError> {
  const parsed = ClaimSchema.safeParse(input);
  return parsed.success
    ? ok(parsed.data)
    : err(
        validationErrorFromZod(
          parsed.error,
          'INVALID_CLAIM',
          'The claim violates the grounding discipline.',
        ),
      );
}

export const UncertaintyKind = {
  /** Reducible: a nameable input or check would resolve it. */
  MISSING_INFORMATION: 'MISSING_INFORMATION',
  /** Irreducible: the outcome is genuinely stochastic; more analysis will not resolve it. */
  STOCHASTIC: 'STOCHASTIC',
} as const;
export const UncertaintyKindSchema = z.enum(['MISSING_INFORMATION', 'STOCHASTIC']);
export type UncertaintyKind = z.infer<typeof UncertaintyKindSchema>;

export const UncertaintySchema = z
  .object({
    kind: UncertaintyKindSchema,
    description: nonEmpty('Uncertainty.description must be non-empty.'),
    /** What would resolve it. Required for MISSING_INFORMATION, forbidden for STOCHASTIC. */
    resolution: z.string().trim().min(1).optional(),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.kind === UncertaintyKind.MISSING_INFORMATION && value.resolution === undefined) {
      ctx.addIssue({
        code: 'custom',
        path: ['resolution'],
        message: 'A MISSING_INFORMATION uncertainty must name what would resolve it.',
      });
    }
    if (value.kind === UncertaintyKind.STOCHASTIC && value.resolution !== undefined) {
      ctx.addIssue({
        code: 'custom',
        path: ['resolution'],
        message: 'A STOCHASTIC uncertainty is irreducible and cannot carry a resolution.',
      });
    }
  });
export type Uncertainty = z.infer<typeof UncertaintySchema>;
export type UncertaintyInput = z.input<typeof UncertaintySchema>;

export function uncertainty(input: UncertaintyInput): Result<Uncertainty, ValidationError> {
  const parsed = UncertaintySchema.safeParse(input);
  return parsed.success
    ? ok(parsed.data)
    : err(
        validationErrorFromZod(
          parsed.error,
          'INVALID_UNCERTAINTY',
          'The uncertainty violates the kind/resolution rules.',
        ),
      );
}
